tests/consensus_spec/test_fixture_kzg.nim

```
@@ -305,6 +308,171 @@ proc runRecoverCellsAndKzgProofsTest(suiteName, suitePath, path: string) =
+
+      # check recovered cells and proofs
+      # assuming columns are sorted
+      var tp = Taskpool.new()
+      let v = tp.recover_cells_and_proofs_parallel(colInput)
+      check v.isOk
+      let val = v.get
+      for i in 0..<val.len:
+        for j in 0..<val[i].cells.len:
+          let k = i * NUMBER_OF_COLUMNS + j
+          check data[k].cell.bytes == val[i].cells[j].bytes
+          check data[k].kzg_proof.bytes == val[i].proofs[j].bytes
+
+proc runRecoverCellsAndKzgProofsParallelInvalidTest(suiteName, suitePath: string) =
+  try:
+    # Preload data of valid cases
+    let
+      validData = loadCellsAndKzgProofsValidCases(suitePath)
+      validCount = validData[validData.len - 1].row_index + 1
+
+    for kind, path in walkDir(suitePath, relative = true, checkDir = true):
+      if not path.contains("_case_invalid_"):
+        continue
+
+      let
+        rowData = loadToJson(os_ops.readFile(suitePath/path/"data.yaml"))[0]
+        inputCells = rowData["input"]["cells"]
+        inputIndices = rowData["input"]["cell_indices"]
+
+      # skip case of different length of inputs as it should have covered in
+      # the single thread tests, and it is not compatible with
+      # `seq[DataColumnSidecar]`
+      if inputCells.len != inputIndices.len:
+        continue
+      # skip case of invalid length
+      if inputCells.len > fulu.CELLS_PER_EXT_BLOB or inputCells.len == 0:
+        continue
+
+      var
+        shouldSkip = false
+        invalidData: seq[MatrixEntry]
+        indices: seq[ColumnIndex]
+
+      for i in 0..<inputCells.len:
+        let
+          cellBytes = fromHex[2048](inputCells[i].getStr).valueOr:
+            # skip case of cell has length != 2048
+            # TODO: support dynamic length for future-proof
+            shouldSkip = true
+            break
+          columnIdx = ColumnIndex(inputIndices[i].getInt.toUInt64.get)
+
+        # Duplicate index is not compatible with `seq[DataColumnSidecar]`
```

How is this the case?

Yes, duplicate indices indicate something wrong and the recovery should fail, but then, part of the point of the invalid tests is to test `recover_cells_and_proofs_parallel`, not to test some ad-hoc filtering code in the test runner. The `seq` itself can have duplicate indices, or indeed, entirely duplicate entries, without issue.

---

I think it is caused by using MatrixEntry to DataColumnSidecar conversion which makes me think it is not compatible. I think I will have to refactor the tests in order to make them more general.